src/anndata/_core/merge.py

```diff
@@ -1399,6 +1404,37 @@ def concat_dataset2d_on_annot_axis(
     return ds_concat_2d
 
 
+def _is_sparse(x):
+    try:
+        return scipy.sparse.issparse(x)
+    except TypeError:
+        return False
+
+
+def _to_numpy_if_array_api(x):
+    if isinstance(x, np.ndarray | pd.DataFrame | pd.Series | DaskArray) or _is_sparse(
+        x
+    ):
+        return x
+    try:
+        import array_api_compat as aac
+
+        # If this succeeds, it's an array-API array (e.g. JAX, cubed, cupy, dask)
+        aac.array_namespace(x)
+        return np.asarray(x)
+    except TypeError:
+        # Not an array-API object (or lib not available) → return unchanged
+        return x
+
+
+def _normalize_nested(obj):
+    if isinstance(obj, dict):
+        return {k: _normalize_nested(v) for k, v in obj.items()}
+    if isinstance(obj, list | tuple):
```

Comment:
Usually you should never do `isinstance(obj, dict)` or `isinstance(obj, list)`, only `Mapping` and `Sequence` (or more precise types). Why did you introduce this?